File: algorithms/searching/quick-select/swift/QuickSelect.swift
class QuickSelect {
    static func select(_ arr: inout [Int], _ k: Int) -> Int {
        return kthSmallest(&arr, 0, arr.count - 1, k)
    }

    private static func kthSmallest(_ arr: inout [Int], _ l: Int, _ r: Int, _ k: Int) -> Int {
        if k > 0 && k <= r - l + 1 {
            let pos = partition(&arr, l, r)

            if pos - l == k - 1 {
                return arr[pos]
            }
            if pos - l > k - 1 {
                return kthSmallest(&arr, l, pos - 1, k)
            }
            return kthSmallest(&arr, pos + 1, r, k - pos + l - 1)
        }
        return -1
    }

    private static func partition(_ arr: inout [Int], _ l: Int, _ r: Int) -> Int {
        let x = arr[r]
        var i = l
        for j in l..<r {
            if arr[j] <= x {
                arr.swapAt(i, j)
                i += 1
            }
        }
        arr.swapAt(i, r)
        return i
    }
}

File: algorithms/searching/quick-select/go/quick_select.go
package quickselect

func QuickSelect(arr []int, k int) int {
	return kthSmallest(arr, 0, len(arr)-1, k)
}

func kthSmallest(arr []int, l, r, k int) int {
	if k > 0 && k <= r-l+1 {
		pos := partition(arr, l, r)

		if pos-l == k-1 {
			return arr[pos]
		}
		if pos-l > k-1 {
			return kthSmallest(arr, l, pos-1, k)
		}
		return kthSmallest(arr, pos+1, r, k-pos+l-1)
	}
	return -1
}

func partition(arr []int, l, r int) int {
	x := arr[r]
	i := l
	for j := l; j <= r-1; j++ {
		if arr[j] <= x {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[r] = arr[r], arr[i]
	return i
}
